content/data/tanzu-gemfire/blog/command-region-pattern.md

```diff
@@ -0,0 +1,163 @@
+---
+title: "The Command Region Pattern"
+description: >
+        The Command Pattern is a behavioural design pattern in which an object is used to encapsulate all information needed to perform an action or trigger an event.
+date: 2020-04-15
+type: blog
+
+# Author(s)
+team: 
+- Juan Jose Ramos
+ 
+---
+
+## Introduction
+[Apache Geode](https://geode.apache.org/) is an in-memory data grid that provides real-time, consistent access to data-intensive applications throughout widely distributed cloud architectures. Its many features include fault-tolerance, high-availability, and WAN replication.
+
+The multi-site or WAN topology is used to connect distinct clusters, they act as one distributed system when they are coupled, and they act as independent systems when communication between sites fails.
+
+The *Command Pattern* is a behavioural design pattern in which an object is used to encapsulate all information needed to perform an action or trigger an event.
+
+
+## Why?
+Gateway Senders are responsible for transmitting events from one site to another and, by default, only the following operations are replicated across WAN sites:
+* Entry put.
+* Entry create.
+* Entry distributed destroy (not originated through expiration).
+
+Sometimes we want to replicate more than that, specifically (but not limited to) notifying the remote sites about *Region Events* (like *destroy* or *clear*) that happened on the local cluster. In order to do so, we can implement the *Command Region Pattern*.
+I’ve seen this pattern successfully and widely applied in a lot of projects over the years but haven’t found a single point of reference about it, now is a good time to share some examples and reasoning behind the pattern itself.
+
+## How?
+The idea is pretty straightforward: a custom *distributedCommand* region will be used for distribution purposes, which will have a *DistributedCommandCacheWriter* attached so we can execute the commands locally. The *DistributedCommand Region* itself will exist on all clusters and the commands will be sent across the wire through Geode gateway-senders so remote clusters can receive them through Geode gateway-receivers and execute the unit of work as well.
+![img](/images/data-blogs/tanzu-gemfire/command-region-pattern/diagrams/command-region-pattern-diagram.jpeg)
+
+1. The client application creates a DistributedCommand instance and executes a put operation through the PROXY region.
+2. The CacheWriter is invoked and the DistributedCommand is executed locally.
+3. The DistributedCommand is added to the gateway-sender queue for distribution to remote clusters.
+4. The DistributedCommand is distributed and received by a gateway-receiver on a remote cluster.
+5. The DistributedCommand reaches a remote server, where the CacheWriter is invoked and the DistributedCommand is executed locally.
+
+## Implementation
+* Create a *DistributedCommand* interface.
```

Comment:
This and all steps under this heading should be formatted as an ordered list. Numbered, as opposed to bulleted.